TombEngine/Objects/TR2/Entity/tr2_sword_guardian.cpp

```diff
@@ -12,210 +13,372 @@
+		SWORD_GUARDIAN_STATE_WAIT = 7,
+		SWORD_GUARDIAN_STATE_FLY = 8,
+		SWORD_GUARDIAN_STATE_AWAKE = 9,
+		SWORD_GUARDIAN_STATE_WALK_ATTACK_AIM = 10,
+		SWORD_GUARDIAN_STATE_WALK_ATTACK = 11,
+		SWORD_GUARDIAN_STATE_DEATH = 12
+	};
+
+	// TODO: Found all name, also found some which is more appropriate.
+	enum SwordGuardianAnim
+	{
+		SWORD_GUARDIAN_ANIM_AWAKE = 0
+	};
+
+	static void SpawnSpearGuardianSmoke(const Vector3i& pos, int roomNumber)
+	{
+		auto& smoke = *GetFreeParticle();
+
+		bool isUnderwater = TestEnvironment(ENV_FLAG_WATER, pos, roomNumber);
+		auto sphere = BoundingSphere(pos.ToVector3(), 16);
+		auto effectPos = Random::GeneratePointInSphere(sphere);
+
+		smoke.on = true;
+		smoke.blendMode = BLEND_MODES::BLENDMODE_ADDITIVE;
+
+		smoke.x = effectPos.x;
+		smoke.y = effectPos.y;
+		smoke.z = effectPos.z;
+		smoke.xVel = Random::GenerateInt(-BLOCK(0.5f), BLOCK(0.5f));
+		smoke.yVel = Random::GenerateInt(-BLOCK(1 / 8.0f), BLOCK(1 / 8.0f));
+		smoke.zVel = Random::GenerateInt(-BLOCK(0.5f), BLOCK(0.5f));
+
+		if (isUnderwater)
+		{
+			smoke.sR = 255;
+			smoke.sG = 255;
+			smoke.sB = 255;
+			smoke.dR = 0;
+			smoke.dG = 0;
+			smoke.dB = 0;
+		}
+		else
+		{
+			smoke.sR = 0;
+			smoke.sG = 0;
+			smoke.sB = 0;
+			smoke.dR = 255;
+			smoke.dG = 255;
+			smoke.dB = 255;
+		}
+
+		smoke.colFadeSpeed = 8;
+		smoke.fadeToBlack = 64;
+		smoke.sLife = smoke.life = Random::GenerateInt(72, 128);
+		smoke.extras = 0;
+		smoke.dynamic = -1;
+
+		if (isUnderwater)
+		{
+			smoke.yVel /= 16;
+			smoke.y += 32;
+			smoke.friction = 4 | 16;
+		}
+		else
+		{
+			smoke.friction = 6;
+		}
+
+		smoke.rotAng = Random::GenerateAngle();
+		smoke.rotAdd = Random::GenerateAngle(ANGLE(-0.2f), ANGLE(0.2f));
+		smoke.flags = SP_SCALE | SP_DEF | SP_ROTATE | SP_EXPDEF;
+		smoke.scalar = 3;
+
+		if (isUnderwater)
+		{
+			smoke.gravity = 0;
+			smoke.maxYvel = 0;
+		}
+		else
+		{
+			smoke.gravity = Random::GenerateInt(-8, -4);
+			smoke.maxYvel = Random::GenerateInt(-8, -4);
+		}
+
+		int scale = Random::GenerateInt(128, 172);
+		smoke.size = smoke.sSize = scale / 8;
+		smoke.dSize = scale;
+	}
+
+	static void SwapMeshToJade(ItemInfo& item, int jointIndex, bool useEffect = true)
+	{
+		if (useEffect)
+			SpawnSpearGuardianSmoke(GetJointPosition(&item, jointIndex), item.RoomNumber);
+		item.Model.MeshIndex[jointIndex] = Objects[ID_SWORD_GUARDIAN_STATUE].meshIndex + jointIndex;
+	}
+
+	static void SwapJadeToNormal(ItemInfo& item, int jointIndex, bool useEffect = true)
+	{
+		if (useEffect)
+			SpawnSpearGuardianSmoke(GetJointPosition(&item, jointIndex), item.RoomNumber);
+		item.Model.MeshIndex[jointIndex] = Objects[ID_SWORD_GUARDIAN].meshIndex + jointIndex;
+	}
+
+	static bool DoSwordGuardianSwapMesh(ItemInfo& item)
+	{
+		const auto& object = Objects[item.ObjectNumber];
+		auto& creature = *GetCreatureInfo(&item);
+
+		// Do mesh swaps.
+		if (creature.Flags == 0)
+		{
+			switch (item.ItemFlags[3])
+			{
+				// Jade to normal.
+			case 0:
```

Comment:
can the numbers be demagiced? :peepohappy:

Reply:
it's what shiva use too, it's not magic, it's an ID which is 0/1

Reply:
Fixed